DAOS-16469 engine: bind helper

Nasf-Fan opened the pull request:
When a main IO XS wants to add some task to helper XS (in spite of via chore queue or creating new ULT), it needs to take lock on related helper XS. Under the dss_helper_pool mode, if there are a lot of ULTs from different main IO XS doing that concurrently, the overhead caused by lock contention may be high. To reduce such overhead, we will divide main IO XS into small groups, and bind each group to a helper XS: the ones in the same group will share one helper, different groups use different helpers. Then above lock contention will be restricted inside the group. Depends on the system configuration, such solution may be unnecessary, for example, there is only single helper per engine, or very limited targets (less than 4 or about) per engine. So this feature will be configurable via environment variable DAOS_BIND_HELPER that is unset by default.

NOTE:

1. Binding helper is only available when the count of helper XS is non-zero and smaller than the count of main IO XS.
2. If multiple sockets mode is enabled (dss_numa_nr > 1), then do not bind helper for simplification.
3. To be load balanced, if the count of main IO XS is not the integral multiple of the helper XS, then binding helper will be disabled automatically.

Allow-unstable-test: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15303/2/execution/node/1211/log

Nasf-Fan committed:
DAOS-16469 engine: bind helper

When a main IO XS wants to add some task to helper XS (in spite of via chore
queue or creating new ULT), it needs to take lock on related helper XS. Under
the dss_helper_pool mode, if there are a lot of ULTs from different main IO XS
doing that concurrently, the overhead caused by lock contention may be high. To
reduce such overhead, we will divide main IO XS into small groups, and bind each
group to a helper XS: the ones in the same group will share one helper, different
groups use different helpers. Then above lock contention will be restricted inside
the group. Depends on the system configuration, such solution may be unnecessary,
for example, there is only single helper per engine, or very limited targets (less
than 4 or about) per engine. So this feature will be configurable via environment
variable DAOS_BIND_HELPER that is unset by default.

NOTE:

1. Binding helper is only available when the count of helper XS is non-zero and
   smaller than the count of main IO XS.
2. If multiple sockets mode is enabled (dss_numa_nr > 1), then do not bind helper
   for simplification.
3. To be load balanced, if the count of main IO XS is not the integral multiple
   of the helper XS, then binding helper will be disabled automatically.

Allow-unstable-test: true

Signed-off-by: Fan Yong <fan.yong@intel.com>
